### PR TITLE
Terminal: Implement Scroll Regions & Scrolling Properly

### PR DESCRIPTION
**Pull request by christopherdumas**
Working on fixing the scrolling and drawing abilities of the Terminal app, so that Vim works properly within it.

### CODE REVIEW

**Comment by christopherdumas** [comment]
This is an *attempt* to fix #136 .

File: Applications/Terminal/Terminal.cpp [hunk 1]
```diff
@@ -373,6 +379,42 @@ void Terminal::escape$J(const ParamVector& params)
...
+
+void Terminal::escape$L(const ParamVector& params)
+{
+    int count = 1;
+    if (params.size() >= 1)
+        count = params[0];
+    dbgprintf("Terminal: Adding %d lines below cursor (at line %d)\n", count, m_cursor_row);
+    invalidate_cursor();
+    for (word row = m_rows; row > m_cursor_row; --row)
+        m_lines[row] = m_lines[row - 1];
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.cpp]: This doesn't look right.
`m_lines[row]` is 1 past the end on the first iteration, since m_lines has only `row` entries.
Also, the lines that we're replacing are raw pointers, and must be deleted, otherwise we leak the memory.

File: Applications/Terminal/Terminal.cpp [hunk 2]
```diff
@@ -388,8 +430,11 @@ void Terminal::escape$M(const ParamVector& params)
     count = min(count, max_count);
 
     dbgprintf("Delete %d line(s) starting from %d\n", count, m_cursor_row);
-    // FIXME: Implement.
-    ASSERT_NOT_REACHED();
+    for (word i = 0; i < count; ++i)
+        delete m_lines[m_cursor_row + i];
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.cpp]: Here we are instead deleting lines but never creating any new ones.
The `m_lines` array must always have `m_rows` valid entries.

File: Applications/Terminal/Terminal.cpp [hunk 3]
```diff
@@ -373,6 +379,42 @@ void Terminal::escape$J(const ParamVector& params)
...
+{
+    int count = 1;
+    if (params.size() >= 1)
+        count = params[0];
+    dbgprintf("Terminal: Adding %d lines below cursor (at line %d)\n", count, m_cursor_row);
+    invalidate_cursor();
+    for (word row = m_rows; row > m_cursor_row; --row)
+        m_lines[row] = m_lines[row - 1];
+    m_lines[m_cursor_row] = new Line(m_columns);
+    ++m_rows_to_scroll_backing_store;
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.cpp]: Shouldn't this be `m_rows_to_scroll_backing_store += count;`?
That variable is for the painting optimization where we know that a chunk of the terminal backing store bitmap can just be shifted vertically upwards to achieve the same result as a normal repaint would.

File: Applications/Terminal/Terminal.cpp [hunk 4]
```diff
@@ -470,14 +518,25 @@ void Terminal::scroll_up()
...
     // NOTE: We have to invalidate the cursor first.
     invalidate_cursor();
-    delete m_lines[0];
-    for (word row = 1; row < rows(); ++row)
+    delete m_lines[m_scroll_region_top];
+    for (word row = m_scroll_region_top + 1; row < m_scroll_region_bottom; ++row)
         m_lines[row - 1] = m_lines[row];
-    m_lines[m_rows - 1] = new Line(m_columns);
+    m_lines[m_scroll_region_bottom - 1] = new Line(m_columns);
     ++m_rows_to_scroll_backing_store;
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.cpp]: I don't think the `m_rows_to_scroll_backing_store` optimization will work correctly if you have a scroll region.
At the very least it probably needs to be aware of it, and take some special action.
Currently it will just move everything up to the first scanline (excluding vertical inset.) If there's a scrolling region set up, that definitely won't do the right thing.

**Comment by christopherdumas** [comment]
This is why I put this up here! The `m_rows_to_scroll_backing_store` misunderstanding was my problem (at least with the bug I'm currently having)! At least, it sounds like it is, I haven't tried it yet.

Re: all the memory leaks and madness that I've created, none of that will be in the final PR. I'm just trying to get something to do something that looks like it might have a delta towards what is the correct behavior at this point. (:

**Comment by awesomekling** [comment]
Oh duh, I keep not noticing when PR's are marked as drafts. That's my bad for thinking this was finished already :)

It looks like you're on the right track, just gotta tame the Terminal repaint code and make it do what you want.

File: AK/Vector.h [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 #include <AK/Assertions.h>
 #include <AK/StdLibExtras.h>
+#include <LibCore/CNotifier.h>
```
Comment (rburchell) [inline AK/Vector.h]: Seems unneeded

File: Applications/Terminal/Terminal.cpp [hunk 6]
```diff
@@ -135,6 +134,36 @@ static inline Color lookup_color(unsigned color)
     return Color::from_rgb(xterm_colors[color]);
 }
 
+void Terminal::escape$l(const ParamVector& params)
+{
+    int mode = 2;
+    if (params.size() > 0) {
+        mode = params[0];
+    }
+    if (!m_question_param) {
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: Consider passing this in as a parameter instead of using a member variable? Less error prone, and more obvious what it's for then.

File: Applications/Terminal/Terminal.cpp [hunk 7]
```diff
@@ -135,6 +134,36 @@ static inline Color lookup_color(unsigned color)
...
+        switch (mode) {
+            // FIXME: implement *something* for this
+        default:
+            unimplemented_escape();
+            break;
+        }
+    } else {
+        switch (mode) {
+        case 25:
+            // Hide cursor command, but doesn't need to be run (for now, because
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: Consider adding a dbgprintf at the least, so it's obvious if someone is looking to implement the functionality later on where they need to edit.

File: Applications/Terminal/Terminal.cpp [hunk 8]
```diff
@@ -135,6 +134,36 @@ static inline Color lookup_color(unsigned color)
...
+            // we don't do inverse control codes anyways)
+            break;
+        default:
+            break;
+        }
+    }
+}
+
+void Terminal::escape$h(const ParamVector& params)
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: IIRC, h and l are basically the same thing. 'h' is to set modes, 'l' is to remove them. It might make sense to combine the two function bodies, and have a boolean 'should_set' or something parameter.

**Comment by christopherdumas** [comment]
@rburchell Thank you for the input! I'll be sure to put it into effect before I make this official.

File: Applications/Terminal/Terminal.cpp (hunk 6 above)
Comment (rburchell) [reply Applications/Terminal/Terminal.cpp]: This doesn't seem to have been done, not sure why it was marked as resolved?

File: Applications/Terminal/Terminal.cpp (hunk 8 above)
Comment (rburchell) [reply Applications/Terminal/Terminal.cpp]: Same here

File: Applications/Terminal/Terminal.cpp [hunk 9]
```diff
@@ -254,7 +303,13 @@ void Terminal::escape$r(const ParamVector& params)
         top = params[0];
     if (params.size() >= 2)
         bottom = params[1];
-    dbgprintf("FIXME: escape$r: Set scrolling region: %u-%u\n", top, bottom);
+    if ((bottom - top) < 2 || bottom > m_rows) {
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: Why 2? I'd guess that a single line region would be valid. Might deserve a comment?

Missing a check for top being too small?

File: Applications/Terminal/Terminal.cpp [hunk 10]
```diff
@@ -265,7 +320,7 @@ void Terminal::escape$H(const ParamVector& params)
         row = params[0];
     if (params.size() >= 2)
         col = params[1];
-    set_cursor(row - 1, col - 1);
+    set_cursor(row - 1 + m_scroll_region_top, col - 1);
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: This should only be done when origin mode is set. It looks like you started adding parsing for this (h/l), but didn't finish?

File: Applications/Terminal/Terminal.cpp [hunk 11]
```diff
@@ -373,7 +428,9 @@ void Terminal::escape$K(const ParamVector& params)
         }
         break;
     case 2:
-        unimplemented_escape();
+        for (int i = 0; i < m_columns; ++i) {
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: Consider adding a comment?

     // Clear the complete line

File: Applications/Terminal/Terminal.cpp [hunk 12]
```diff
@@ -389,18 +446,22 @@ void Terminal::escape$J(const ParamVector& params)
...
         for (int row = m_cursor_row + 1; row < m_rows; ++row) {
             for (int column = 0; column < m_columns; ++column) {
                 put_character_at(row, column, ' ');
             }
         }
         break;
     case 1:
-        // FIXME: Clear from cursor to beginning of screen.
-        unimplemented_escape();
+        for (int i = m_cursor_column-1; i >= 0; --i)
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: I'd suggest leaving the comment in place:

    // Clear from the beginning of the display to the cursor

File: Applications/Terminal/Terminal.cpp [hunk 13]
```diff
@@ -389,18 +446,22 @@ void Terminal::escape$J(const ParamVector& params)
...
+            put_character_at(m_cursor_row, i, ' ');
+        for (int row = m_cursor_row-2; row >= 0; --row) {
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: Hmm, why is this m_cursor_row-2? We just cleared m_cursor_row above, so aren't we then missing some rows?

File: Applications/Terminal/Terminal.cpp [hunk 14]
```diff
@@ -415,6 +476,42 @@ void Terminal::escape$J(const ParamVector& params)
...
+{
+    int count = 1;
+    if (params.size() >= 1)
+        count = params[0];
+
+    for (word i = 0; i < count; i++)
+        scroll_up();
+}
+
+void Terminal::escape$T(const ParamVector& params)
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: Might be useful to add comments pointing to what part of the VT docs these are (SD for T, SU for S)

(Just a general comment, as it makes finding them to check what they are supposed to do easier)

File: Applications/Terminal/Terminal.cpp [hunk 15]
```diff
@@ -457,20 +558,34 @@ void Terminal::execute_xterm_command()
...
+            // FIXME: Should we do something else?
             m_parameters.clear_with_capacity();
             m_intermediates.clear_with_capacity();
-            // FIXME: Should we do something else?
             return;
         }
         params.append(value);
     }
+
+    dbgprintf("Terminal::execute_escape_sequence: Handled final '%c'\n", final);
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: This is going to be pretty noisy. I'd suggest sticking it in an:

    #if defined(TERMINAL_DEBUG)
    #endif

File: Applications/Terminal/Terminal.cpp [hunk 16]
```diff
@@ -488,19 +606,30 @@ void Terminal::execute_escape_sequence(byte final)
...
     case 't': escape$t(params); break;
     case 'r': escape$r(params); break;
+    case 'l': escape$l(params); break;
+    case 'h': escape$h(params); break;
     default:
         dbgprintf("Terminal::execute_escape_sequence: Unhandled final '%c'\n", final);
         break;
     }
 
+    dbgprintf("\n");
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: Same here (#if defined(TERMINAL_DEBUG))

File: Applications/Terminal/Terminal.cpp [hunk 17]
```diff
@@ -709,15 +842,20 @@ void Terminal::set_size(word columns, word rows)
     if (columns == m_columns && rows == m_rows)
         return;
 
-    if (m_lines) {
-        for (size_t i = 0; i < m_rows; ++i)
-            delete m_lines[i];
-        delete m_lines;
+    dbgprintf("Terminal: RESIZE to: %d rows\n", rows);
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: #if defined(TERMINAL_DEBUG)

File: Applications/Terminal/Terminal.cpp [hunk 18]
```diff
@@ -825,6 +958,8 @@ void Terminal::keydown_event(GKeyEvent& event)
     case KeyCode::Key_End:
         write(m_ptm_fd, "\033[F", 3);
         break;
+    case KeyCode::Key_RightShift:
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: Hmm, the purpose of this isn't obvious to me at least. Maybe add a comment explaining the reason for it being ignored.

File: Applications/Terminal/Terminal.cpp [hunk 19]
```diff
@@ -837,52 +972,37 @@ void Terminal::paint_event(GPaintEvent& event)
...
-        memcpy(
-            painter.target()->scanline(first_scanline),
-            painter.target()->scanline(second_scanline),
-            scanlines_to_copy * painter.target()->pitch()
-        );
-        line(max(0, m_cursor_row - m_rows_to_scroll_backing_store)).dirty = true;
-    }
-    m_rows_to_scroll_backing_store = 0;
-
+    if (m_visual_beep_timer.is_active())
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: Hrm, full red is going to be a little bit retina-melting. Maybe it should be a dark red, or something? Probably OK at least for now.

File: Applications/Terminal/Terminal.cpp [hunk 20]
```diff
@@ -837,52 +972,37 @@ void Terminal::paint_event(GPaintEvent& event)
...
-            continue;
-        line.dirty = false;
         bool has_only_one_background_color = line.has_only_one_background_color();
         if (m_visual_beep_timer.is_active())
             painter.fill_rect(row_rect(row), Color::Red);
         else if (has_only_one_background_color)
             painter.fill_rect(row_rect(row), lookup_color(line.attributes[0].background_color).with_alpha(255 * m_opacity));
         for (word column = 0; column < m_columns; ++column) {
-            bool should_reverse_fill_for_cursor = m_cursor_blink_state && m_in_active_window && row == m_cursor_row && column == m_cursor_column;
+            bool should_reverse_fill_for_cursor = m_cursor_blink_state
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: No need to reformat this (just stick to what clang-format wants to do formatting-wise)

File: Applications/Terminal/Terminal.cpp [hunk 21]
```diff
@@ -837,52 +972,37 @@ void Terminal::paint_event(GPaintEvent& event)
...
             auto character_rect = glyph_rect(row, column);
             if (!has_only_one_background_color || should_reverse_fill_for_cursor) {
                 auto cell_rect = character_rect.inflated(0, m_line_spacing);
-                painter.fill_rect(cell_rect, lookup_color(should_reverse_fill_for_cursor ? attribute.foreground_color : attribute.background_color).with_alpha(255 * m_opacity));
+                painter.fill_rect(cell_rect, lookup_color(should_reverse_fill_for_cursor
+                                                          ? attribute.foreground_color
+                                                          : attribute.background_color)
+                                  .with_alpha(255 * m_opacity));
             }
-            if (ch == ' ')
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: Why was this removed?

File: Applications/Terminal/Terminal.cpp [hunk 22]
```diff
@@ -837,52 +972,37 @@ void Terminal::paint_event(GPaintEvent& event)
...
 
-    if (m_needs_background_fill) {
-        m_needs_background_fill = false;
-        if (m_visual_beep_timer.is_active())
-            painter.fill_rect(frame_inner_rect(), Color::Red);
-        else
-            painter.fill_rect(frame_inner_rect(), Color(Color::Black).with_alpha(255 * m_opacity));
-    }
-
-    if (m_rows_to_scroll_backing_store && m_rows_to_scroll_backing_store < m_rows) {
```
Comment (rburchell) [inline Applications/Terminal/Terminal.cpp]: Did you establish whether or not there was actually any problem with the repaint optimization after adding the missing escape handling? It would be a shame to lose it unnecessarily.

File: Applications/Terminal/Terminal.h [hunk 23]
```diff
@@ -164,6 +173,7 @@ class Terminal final : public GFrame {
     };
     EscapeState m_escape_state { Normal };
     Vector<byte> m_parameters;
+    bool m_question_param;
```
Comment (rburchell) [inline Applications/Terminal/Terminal.h]: As mentioned above I think, I'd prefer to see this removed as a member (and instead, passed as a parameter).

File: Kernel/.gitignore [hunk 24]
```diff
@@ -7,3 +7,4 @@ _fs_contents
 sync-local.sh
 *.pcap
 eth_null*
+compile_commmands.json
```
Comment (rburchell) [inline Kernel/.gitignore]: What's this? Well, OK if it makes your life easier I suppose, but it seems a little arbitrary

File: Kernel/makeall.sh [hunk 25]
```diff
@@ -7,7 +7,7 @@ if [ -z "$MAKEJOBS" ]; then
     MAKEJOBS=$(nproc)
 fi
 
-make_cmd="make -j $MAKEJOBS"
+make_cmd="bear make -j $MAKEJOBS"
```
Comment (rburchell) [inline Kernel/makeall.sh]: This seems very wrong

File: LibC/termcap.cpp [hunk 26]
```diff
@@ -115,7 +115,7 @@ char* tgoto(const char* cap, int col, int row)
     (void)cap;
     (void)col;
     (void)row;
-    ASSERT_NOT_REACHED();
+    //ASSERT_NOT_REACHED();
```
Comment (rburchell) [inline LibC/termcap.cpp]: Just removing this seems wrong. It was before you were using vim's own termcap stuff, wasn't it?

File: LibC/unistd.h [hunk 27]
```diff
@@ -109,4 +109,7 @@ enum
  */
 #define _POSIX_PRIORITY_SCHEDULING
 
+// Stifle an less error iirc
+#define _PC_VDISABLE 8
```
Comment (rburchell) [inline LibC/unistd.h]: Remove the comment, and move this to sit in the enum with _PC_NAME_MAX in the same file (since they are used in the same function) {f}pathconf.

File: Applications/Terminal/Terminal.cpp (hunk 9 above)
Comment (christopherdumas) [reply Applications/Terminal/Terminal.cpp]: So according to the reference material I have, scroll regions are not allowed to have less than two rows. Your guess is as good as mine concerning why...

**Comment by christopherdumas** [comment]
> My earlier comments still seem valid, also a few small fixes and questions.
I thought I got all of your comments from earlier (I hope my changes didn't get clobbered somehow...).
> I think the only real functional question I have is whether or not removing the repaint optimization is still necessary, it would be a shame to lose that.
Honestly, I don't know, I haven't tried to add them back in yet... I thought I would make another PR for that, since the goal of this PR is honestly just to get stuff working.

File: Applications/Terminal/Terminal.cpp (hunk 19 above)
Comment (christopherdumas) [reply Applications/Terminal/Terminal.cpp]: Yeah that makes sense lol, but we already merged the PR that added/uses this color. Take it up with @awesomekling if you want a different design decision, he seemed happy with it.

File: Applications/Terminal/Terminal.cpp (hunk 21 above)
Comment (christopherdumas) [reply Applications/Terminal/Terminal.cpp]: So either we don't repaint the background every time, or we remove this. I guess I'll add that back in for now...

File: Kernel/.gitignore (hunk 24 above)
Comment (christopherdumas) [reply Kernel/.gitignore]: > What's this?

It's essentially a project file for Emacs, so I get completion, etc. Like the VS Code stuff I've seen in .gitignores in other places.

> Well, OK if it makes your life easier I suppose, but it seems a little arbitrary

I'm not sure how to force exclude a file from git while also being able to keep it between commits and such, without adding it to the gitignore.

File: Kernel/makeall.sh (hunk 25 above)
Comment (christopherdumas) [reply Kernel/makeall.sh]: Ah, that wasn't supposed to make it in...

File: LibC/termcap.cpp (hunk 26 above)
Comment (christopherdumas) [reply LibC/termcap.cpp]: No, this prevents Bash and Less from crashing randomly.

File: LibC/unistd.h (hunk 27 above)
Comment (christopherdumas) [reply LibC/unistd.h]: Gotcha.

**Comment by christopherdumas** [comment]
> I think the only real functional question I have is whether or not removing the repaint optimization is still necessary, it would be a shame to lose that.

I did some testing and experiments, and essentially, yes, the backing store and dirty lines implementations were *not* compatible with any kind of scrolling more complex than one line up/one line down, and especially not scroll regions. @awesomekling OKed a first PR w/o them, and in my next PR I'll do my best to get versions of them working with the new scroll regions. The backing store will have to be much more complicated than just "add a row", but it could be possible to get working. Same for dirty lines. But it didn't seem like it should be part of this PR.

File: LibC/termcap.cpp (hunk 26 above)
Comment (rburchell) [reply LibC/termcap.cpp]: Then it sounds like it belongs in a separate PR, and it should also not just be commented out IMO, as it's definitely not correct to just stub functions without at least making it clear that they _are_ stubs, as then when trying to figure out why behaviour is incorrect, you have a much harder job as you don't know where to start looking :)

Either the function should be implemented, or at the least, it should be turned into a dbgprintf if the assert is to be removed -- but I'd say that leaving it as an assert is the safer option so it's clear that there is still work to do.

File: Applications/Terminal/Terminal.cpp (hunk 19 above)
Comment (rburchell) [reply Applications/Terminal/Terminal.cpp]: Right, I didn't see that it was already there.

Comment (awesomekling) [reply Applications/Terminal/Terminal.cpp]: Yeah, the red is definitely a much and we should improve on it. Let's keep the scope of each PR tight though.

Comment (rburchell) [reply Applications/Terminal/Terminal.cpp]: Yep, I wouldn't have commented if I had read the left hand side of the diff closely. I didn't realise it had already been added, it was a nice new feature to me. ;-)

**Comment by christopherdumas** [comment]
@rburchell Okay I made those changes. Sorry for not doing them originally. Also, we are using cargo-format now, correct? In this last commit I enabled it and ran it on my code, which causes a lot of minor formatting changes.

**Comment by awesomekling** [comment]
I don't understand the state of this PR, is it ready for review?
Also , "rework structure a bit, and reformat" is very muddy, please don't mix automated reformatting commits changes with actual logic changes. It's okay to run `clang-format` on a file to tidy it up, but please put that in a separate commit :)

**Comment by christopherdumas** [comment]
It's ready for review. Also, yeah the format was accidental, so I wasn't really able to separate it. ): After it happened I was able to figure out how to prevent accidental formats from happening.

File: Applications/Terminal/Terminal.cpp (hunk 14 above)
Comment (awesomekling) [reply Applications/Terminal/Terminal.cpp]: I think the best thing would be getting rid of the `escape$xxx` names and replacing them with names from the VT docs instead. :)
Not for this PR, but a nice future improvement.

File: Applications/Terminal/Terminal.cpp [hunk 28]
```diff
@@ -415,6 +455,42 @@ void Terminal::escape$J(const ParamVector& params)
...
+}
+
+void Terminal::escape$L(const ParamVector& params)
+{
+    int count = 1;
+    if (params.size() >= 1)
+        count = params[0];
+    invalidate_cursor();
+    for (; count > 0; --count) {
+        m_lines.insert(m_cursor_row + m_scroll_region_top, OwnPtr(new Line(m_columns)));
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.cpp]:
```suggestion
        m_lines.insert(m_cursor_row + m_scroll_region_top, make<Line>(m_columns));

File: Applications/Terminal/Terminal.cpp [hunk 29]
```diff
@@ -415,6 +455,42 @@ void Terminal::escape$J(const ParamVector& params)
...
+        if (m_scroll_region_bottom < m_lines.size())
+            m_lines.remove(m_scroll_region_bottom + 1);
+        else
+            m_lines.remove(m_lines.size());
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.cpp]: This won't work, `Vector::remove()` will assert that `index < m_size`.
To remove the last entry, use `m_lines.remove(m_lines.size() - 1);`
It might be nice to add a `Vector::remove_last()` helper specifically for this.

File: Applications/Terminal/Terminal.cpp [hunk 30]
```diff
@@ -426,12 +502,16 @@ void Terminal::escape$M(const ParamVector& params)
...
 
-    dbgprintf("Delete %d line(s) starting from %d\n", count, m_cursor_row);
-    // FIXME: Implement.
-    ASSERT_NOT_REACHED();
+    for (int c = count; c > 0; --c) {
+        m_lines.remove(m_cursor_row + m_scroll_region_top);
+        if (m_scroll_region_bottom < m_lines.size())
+            m_lines.insert(m_scroll_region_bottom, OwnPtr(new Line(m_columns)));
+        else
+            m_lines.append(OwnPtr(new Line(m_columns)));
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.cpp]:
```suggestion
            m_lines.append(make<Line>(m_columns));
```

File: Applications/Terminal/Terminal.cpp [hunk 31]
```diff
@@ -426,12 +502,16 @@ void Terminal::escape$M(const ParamVector& params)
...
+    int max_count = m_rows - (m_scroll_region_top + m_cursor_row);
     count = min(count, max_count);
 
-    dbgprintf("Delete %d line(s) starting from %d\n", count, m_cursor_row);
-    // FIXME: Implement.
-    ASSERT_NOT_REACHED();
+    for (int c = count; c > 0; --c) {
+        m_lines.remove(m_cursor_row + m_scroll_region_top);
+        if (m_scroll_region_bottom < m_lines.size())
+            m_lines.insert(m_scroll_region_bottom, OwnPtr(new Line(m_columns)));
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.cpp]:
```suggestion
            m_lines.insert(m_scroll_region_bottom, make<Line>(m_columns));
```

File: Applications/Terminal/Terminal.cpp [hunk 32]
```diff
@@ -512,11 +666,17 @@ void Terminal::scroll_up()
 {
     // NOTE: We have to invalidate the cursor first.
     invalidate_cursor();
-    delete m_lines[0];
-    for (word row = 1; row < rows(); ++row)
-        m_lines[row - 1] = m_lines[row];
-    m_lines[m_rows - 1] = new Line(m_columns);
-    ++m_rows_to_scroll_backing_store;
+    m_lines.remove(m_scroll_region_top);
+    m_lines.insert(m_scroll_region_bottom, OwnPtr(new Line(m_columns)));
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.cpp]:
```suggestion
    m_lines.insert(m_scroll_region_bottom, make<Line>(m_columns));
```

File: Applications/Terminal/Terminal.cpp [hunk 33]
```diff
@@ -512,11 +666,17 @@ void Terminal::scroll_up()
...
+    m_need_full_flush = true;
+}
+
+void Terminal::scroll_down()
+{
+    // NOTE: We have to invalidate the cursor first.
+    invalidate_cursor();
+    m_lines.remove(m_scroll_region_bottom);
+    m_lines.insert(m_scroll_region_top, OwnPtr(new Line(m_columns)));
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.cpp]:
```suggestion
    m_lines.insert(m_scroll_region_top, make<Line>(m_columns));
```

File: Applications/Terminal/Terminal.cpp [hunk 34]
```diff
@@ -709,15 +867,23 @@ void Terminal::set_size(word columns, word rows)
...
-        for (size_t i = 0; i < m_rows; ++i)
-            delete m_lines[i];
-        delete m_lines;
+#if defined(TERMINAL_DEBUG)
+    dbgprintf("Terminal: RESIZE to: %d rows\n", rows);
+#endif
+
+    if (rows > m_rows) {
+        while (m_lines.size() < rows)
+            m_lines.append(OwnPtr(new Line(columns)));
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.cpp]:
```suggestion
             m_lines.append(make<Line>(columns));
```

File: Applications/Terminal/Terminal.h [hunk 35]
```diff
@@ -132,7 +138,10 @@ class Terminal final : public GFrame {
         return *m_lines[index];
     }
 
-    Line** m_lines { nullptr };
+    Vector<OwnPtr<Line>> m_lines {};
```
Comment (awesomekling) [inline Applications/Terminal/Terminal.h]: No need for {} here.
```suggestion
    Vector<OwnPtr<Line>> m_lines;
```

File: LibC/unistd.h [hunk 36]
```diff
@@ -94,6 +94,7 @@ int ftruncate(int fd, off_t length);
 enum
 {
     _PC_NAME_MAX,
+    _PC_VDISABLE = 8,
```
Comment (awesomekling) [inline LibC/unistd.h]: This should be in a commit by itself with an explanation of what it's for.

**Comment by christopherdumas** [comment]
Okay, I made the requested changes.

**Comment by awesomekling** [comment]
Okay, let's merge it!
